src/routes/authentication.ts

```diff
@@ -30,28 +38,36 @@ export default async function authentication(fastify: FastifyInstance) {
     const data: any = req.body;
     const { username, password, domain } = data;
 
-    const authInfo = Config.getAuthenticationInfo(domain);
-    let chtSession;
     try {
-      chtSession = await ChtSession.create(authInfo, username, password);
-    } catch (e: any) {
-      console.error(`Login error: ${e}`);
-      return resp.view('src/liquid/auth/authentication_form.html', {
-        domains: Config.getDomains(),
-        errors: true,
+      const authInfo = Config.getAuthenticationInfo(domain);
+
+      if (!username || !password) {
+        throw AuthErrors.MISSING_CREDENTIALS();
+      }
+
+      const chtSession = await ChtSession.create(authInfo, username, password);
+      const tokenizedSession = Auth.encodeTokenForCookie(chtSession);
+
+      resp.setCookie(Auth.AUTH_COOKIE_NAME, tokenizedSession, {
+        signed: false,
+        httpOnly: true,
+        expires: Auth.cookieExpiry(),
+        secure: true
       });
-    }
 
-    const tokenizedSession = Auth.encodeTokenForCookie(chtSession);
-    const expires = Auth.cookieExpiry();
-    resp.setCookie(Auth.AUTH_COOKIE_NAME, tokenizedSession, {
-      signed: false,
-      httpOnly: true,
-      expires,
-      secure: true
-    });
+      resp.header('HX-Redirect', '/');
+    } catch (e: any ) {
+      if (e instanceof AuthError) {
```

Comment:
suggest a new function `getLoginErrorMessage` which returns the error message? one call to render